### PR TITLE
feature: generate_presigned_url

### PR DESCRIPTION
**Pull request by DyfanJones**
Add support for `generate_presigned_url`. This gives paws similar features to boto3 (https://docs.aws.amazon.com/AmazonS3/latest/userguide/ShareObjectPreSignedURL.html):

# Boto3

```python
import boto3
url = boto3.client('s3').generate_presigned_url(
    ClientMethod='get_object', 
    Params={'Bucket': 'BUCKET_NAME', 'Key': 'OBJECT_KEY'},
    ExpiresIn=3600
)
```

# Paws
```r
url = paws::s3()$generate_presigned_url(
  ClientMethod='get_object', 
  Params=list(Bucket='BUCKET_NAME', Key='OBJECT_KEY'),
  ExpiresIn=3600
)
```
Issue: https://github.com/paws-r/paws/issues/572

### CODE REVIEW

File: paws.common/R/generate_presigned_url.R [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+#' @param expires_in: The number of seconds the presigned url is valid
+#' for. By default it expires in an hour (3600 seconds)
+#' @param  http_param The http method to use on the generated url. By default,
+#' the http method is whatever is used in the method's model.
+#' @return The presigned url character
+generate_presigned_url <- function(client_method,
+                                params = list(),
+                                expires_in = 3600,
+                                http_method = NULL) {
+  stopifnot(
```
Comment (davidkretch) [inline paws.common/R/generate_presigned_url.R:21]: Nice.

File: paws.common/R/generate_presigned_url.R [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  # build request
+  request <- build(request)
+  # sign request
+  request <- sign_v1_auth_query(request)
+
+  if (!is.null(http_method)) {
+    request$http_request$url$scheme <- http_method
+  }
+
+  return(build_url(request$http_request$url))
```
Comment (davidkretch) [inline paws.common/R/generate_presigned_url.R:102]: Wow.

**Comment by davidkretch** [comment]
Looks good to me, but looks like GitHub Actions isn't happy with the metaprogramming solution used here. My guess is we need to tell it to install paws when running tests for paws.common.

**Comment by DyfanJones** [comment]
Aye I forget about it when I was testing locally :P I might move the `generate_presigned_url` function to the `make.paws` package so that we don't have any issues with cran.  I can have the `make.paws` package have a suggestion to the `paws.storage` so that it can do the unit tests and that :)

File: make.paws/tests/testthat/test_generate_presigned_url.R [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  actual <- svc$generate_presigned_url(
+    client_method = "list_objects_v2",
+    params = list(Bucket = "foo", Prefix = "bar"),
+    http_method = "HTTP"
+  )
+
+  expect_true(
+    grepl(
+      sprintf(
+        "HTTP://%s.s3.amazonaws.com/\\?list-type=2&prefix=%s&AWSAccessKeyId=%s&Expires=.*?&Signature=.*",
```
Comment (davidkretch) [inline make.paws/tests/testthat/test_generate_presigned_url.R:28]: Nice.

**Comment by davidkretch** [comment]
Looks good!

**Pull request merged**